// @raycast/eslint-plugin/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const prefer_ellipsis_1 = __importDefault(require("./rules/prefer-ellipsis"));
const prefer_placeholders_1 = __importDefault(require("./rules/prefer-placeholders"));
const prefer_title_case_1 = __importDefault(require("./rules/prefer-title-case"));
const prefer_common_shortcut_1 = __importDefault(require("./rules/prefer-common-shortcut"));
const no_reserved_shortcut_1 = __importDefault(require("./rules/no-reserved-shortcut"));
const no_ambiguous_platform_shortcut_1 = __importDefault(require("./rules/no-ambiguous-platform-shortcut"));
const pkg = JSON.parse(fs_1.default.readFileSync(path_1.default.join(__dirname, "..", "package.json"), "utf8"));
const plugin = {
    meta: {
        name: pkg.name,
        version: pkg.version,
    },
    configs: {},
    rules: {
        "prefer-ellipsis": prefer_ellipsis_1.default,
        "prefer-title-case": prefer_title_case_1.default,
        "prefer-placeholders": prefer_placeholders_1.default,
        "prefer-common-shortcut": prefer_common_shortcut_1.default,
        "no-reserved-shortcut": no_reserved_shortcut_1.default,
        "no-ambiguous-platform-shortcut": no_ambiguous_platform_shortcut_1.default,
    },
};
Object.assign(plugin.configs, {
    recommended: [
        {
            plugins: {
                "@raycast": plugin,
            },
            rules: {
                "@raycast/prefer-ellipsis": "warn",
                "@raycast/prefer-title-case": "warn",
                "@raycast/prefer-common-shortcut": "warn",
                "@raycast/no-reserved-shortcut": "warn",
                "@raycast/no-ambiguous-platform-shortcut": "warn",
            },
        },
    ],
});
module.exports = plugin;
