// @raycast/eslint-plugin/dist/rules/no-ambiguous-platform-shortcut.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const utils_1 = require("@typescript-eslint/utils");
const utils_2 = require("../utils");
const packagePlatformsCache = new Map();
function readPackagePlatformsConfig(pkgPath) {
    let stats;
    try {
        stats = fs_1.default.statSync(pkgPath);
    }
    catch (_a) {
        packagePlatformsCache.delete(pkgPath);
        return undefined;
    }
    if (!stats.isFile()) {
        packagePlatformsCache.delete(pkgPath);
        return undefined;
    }
    const cached = packagePlatformsCache.get(pkgPath);
    if (cached &&
        cached.mtimeMs === stats.mtimeMs &&
        cached.size === stats.size) {
        return cached.hasMultiPlatform;
    }
    try {
        const pkg = JSON.parse(fs_1.default.readFileSync(pkgPath, "utf8"));
        const platforms = pkg === null || pkg === void 0 ? void 0 : pkg.platforms;
        const hasMultiPlatform = Array.isArray(platforms) && platforms.length > 1;
        packagePlatformsCache.set(pkgPath, {
            hasMultiPlatform,
            mtimeMs: stats.mtimeMs,
            size: stats.size,
        });
        return hasMultiPlatform;
    }
    catch (_b) {
        packagePlatformsCache.set(pkgPath, {
            hasMultiPlatform: false,
            mtimeMs: stats.mtimeMs,
            size: stats.size,
        });
        return false;
    }
}
function hasMultiPlatformConfig(filename) {
    if (!filename || filename.startsWith("<")) {
        return false;
    }
    let dir = path_1.default.dirname(filename);
    while (true) {
        const pkgPath = path_1.default.join(dir, "package.json");
        const hasMultiPlatform = readPackagePlatformsConfig(pkgPath);
        if (hasMultiPlatform !== undefined) {
            return hasMultiPlatform;
        }
        const parent = path_1.default.dirname(dir);
        if (parent === dir) {
            break;
        }
        dir = parent;
    }
    return false;
}
function parseSimpleShortcut(node) {
    const props = new Map();
    for (const prop of node.properties) {
        if (prop.type !== utils_1.AST_NODE_TYPES.Property ||
            prop.key.type !== utils_1.AST_NODE_TYPES.Identifier) {
            return null;
        }
        if (prop.key.name !== "modifiers" && prop.key.name !== "key") {
            return null;
        }
        props.set(prop.key.name, prop);
    }
    if (props.size !== 2) {
        return null;
    }
    const modifiersProp = props.get("modifiers");
    const keyProp = props.get("key");
    if (!modifiersProp || !keyProp)
        return null;
    if (modifiersProp.value.type !== utils_1.AST_NODE_TYPES.ArrayExpression ||
        keyProp.value.type !== utils_1.AST_NODE_TYPES.Literal ||
        typeof keyProp.value.value !== "string") {
        return null;
    }
    const modifiers = [];
    for (const el of modifiersProp.value.elements) {
        if (!el)
            continue;
        if (el.type !== utils_1.AST_NODE_TYPES.Literal || typeof el.value !== "string") {
            return null;
        }
        modifiers.push(el.value);
    }
    return { modifiers, key: keyProp.value.value };
}
function isAmbiguous(modifiers) {
    const hasCmd = modifiers.includes("cmd");
    const hasCtrl = modifiers.includes("ctrl");
    return (hasCmd || hasCtrl) && !(hasCmd && hasCtrl);
}
exports.default = (0, utils_2.createRule)({
    name: "no-ambiguous-platform-shortcut",
    meta: {
        type: "problem",
        docs: {
            description: "Warn when a shortcut is ambiguous in cross-platform extensions.",
        },
        schema: [],
        messages: {
            ambiguous: "This shortcut is ambiguous across platforms. Provide platform-specific shortcuts.",
        },
    },
    defaultOptions: [],
    create(context) {
        const hasMultiPlatform = hasMultiPlatformConfig(context.filename);
        if (!hasMultiPlatform) {
            return {};
        }
        return {
            JSXAttribute(node) {
                if (node.name.type === utils_1.AST_NODE_TYPES.JSXIdentifier &&
                    node.name.name === "shortcut" &&
                    node.value &&
                    node.value.type === utils_1.AST_NODE_TYPES.JSXExpressionContainer &&
                    node.value.expression.type === utils_1.AST_NODE_TYPES.ObjectExpression) {
                    const simple = parseSimpleShortcut(node.value.expression);
                    if (!simple)
                        return;
                    if (!isAmbiguous(simple.modifiers))
                        return;
                    context.report({
                        node: node.value,
                        messageId: "ambiguous",
                    });
                }
            },
        };
    },
});
